analyzers/src/SonarAnalyzer.CFG/Roslyn/CfgAllPathValidator.cs

```diff
@@ -23,29 +23,43 @@ namespace SonarAnalyzer.CFG.Roslyn
     public abstract class CfgAllPathValidator
     {
         private readonly ControlFlowGraph cfg;
-        private readonly Dictionary<BasicBlock, bool> visitedStatus = new Dictionary<BasicBlock, bool>();
 
         protected abstract bool IsValid(BasicBlock block);
         protected abstract bool IsInvalid(BasicBlock block);
 
         protected CfgAllPathValidator(ControlFlowGraph cfg) =>
             this.cfg = cfg;
 
-        public bool CheckAllPaths() =>
-            IsBlockOrAllSuccessorsValid(cfg.EntryBlock);
-
-        private bool IsBlockOrAllSuccessorsValid(BasicBlock block)
-        {
-            var isValid = !IsInvalid(block) && (IsValid(block) || AreAllSuccessorsValid(block));
-            visitedStatus[block] = isValid;
-            return isValid;
-        }
-
-        private bool AreAllSuccessorsValid(BasicBlock block)
+        public bool CheckAllPaths()
         {
-            visitedStatus[block] = true; // protects from loops, don't fail the computation if hits itself
-            return block.SuccessorBlocks.Any()
-                   && block.SuccessorBlocks.All(x => x != cfg.ExitBlock && (visitedStatus.ContainsKey(x) ? visitedStatus[x] : IsBlockOrAllSuccessorsValid(x)));
+            HashSet<BasicBlock> visited = [];
+            var blocks = new Stack<BasicBlock>();
+            blocks.Push(cfg.EntryBlock);
+            while (blocks.Count > 0)
+            {
+                var block = blocks.Pop();
+                if (!visited.Add(block))
+                {
+                    continue; // We already visited this block. (This protects from endless loops)
+                }
+                if (IsInvalid(block) || block == cfg.ExitBlock)
```

Comment:
I would swap this, because ExitBlock test is cheaper and we don't need to invoke IsInvalid computation for those
```suggestion
                if (cfg.ExitBlock || IsInvalid(block))
```